### PR TITLE
router: fix dynamic weight crash

### PR DESCRIPTION
**Pull request by wdauchy**
Commit Message:

fix up crash while updating weighted clusters, when the total weight is changing through runtime keys.

Additional Description:

```
[assert] [source/common/router/config_impl.cc:1366] panic: unexpected
[backtrace] [./source/server/backtrace.h:127] Caught Aborted, suspect faulting address 0xfffe00000001
[backtrace] [./source/server/backtrace.h:111] Backtrace (use tools/stack_decode.py to get line numbers):
[backtrace] [./source/server/backtrace.h:112] Envoy version: 16a0cedaba5335a8003c5077a07a49fafd3eea97/1.30.1/Clean/RELEASE/BoringSSL
[backtrace] [./source/server/backtrace.h:114] Address mapping: 560c90beb000-560c93592000 /usr/local/bin/envoy
[backtrace] [./source/server/backtrace.h:121] #0: [0x7f7cc0a1f520]
ActiveStream 0x55d99505500, stream_id_: 7019320575973121232&filter_manager_:
  FilterManager 0x55d995055b0, state_.has_1xx_headers_: 0
  filter_manager_callbacks_.requestHeaders():
  &streamInfo():
    StreamInfoImpl 0x55d995056f0, protocol_: 1, response_code_: null, response_code_details_: null, attempt_count_: null, health_check_request_: 0, getRouteName():     upstream_info_:     null
    OverridableRemoteConnectionInfoSetterStreamInfo 0x55d995056f0, remoteAddress(): 10.128.7.99:18710, directRemoteAddress(): 10.128.7.99:18710, localAddress(): 10.128.38.140:6032
Http1::ConnectionImpl 0x55daa38f810, dispatching_: 1, dispatching_slice_already_drained_: 0, reset_stream_called_: 0, handling_upgrade_: 0, deferred_end_stream_headers_: 0, processing_trailers_: 0, buffered_body_.length(): 0, header_parsing_state_: Field, current_header_field_: , current_header_value_:
active_request_:
, request_url_: null, response_encoder_.local_end_stream_: 0
absl::get<RequestHeaderMapPtr>(headers_or_trailers_): null
ConnectionImpl 0x55da6d047d0, connecting_: 0, bind_error_: 0, state(): Open, read_buffer_limit_: 131072
socket_:
  ListenSocketImpl 0x55db7be3ef0, transport_protocol_: raw_buffer
  connection_info_provider_:
    ConnectionInfoSetterImpl 0x55da4b54b18, remote_address_: 10.128.7.99:18710, direct_remote_address_: 10.128.7.99:18710, local_address_: 10.128.38.140:6032, server_name_:
[backtrace] [./source/server/backtrace.h:127] Caught Segmentation fault, suspect faulting address 0x0
[backtrace] [./source/server/backtrace.h:111] Backtrace (use tools/stack_decode.py to get line numbers):
[backtrace] [./source/server/backtrace.h:112] Envoy version: 16a0cedaba5335a8003c5077a07a49fafd3eea97/1.30.1/Clean/RELEASE/BoringSSL
[backtrace] [./source/server/backtrace.h:114] Address mapping: 560c90beb000-560c93592000 /usr/local/bin/envoy
[backtrace] [./source/server/backtrace.h:121] #0: [0x7f7cc0a1f520]
Our FatalActions triggered a fatal signal.
```

Risk Level: Low

Testing:

We update the tests to reflect the changes which are now taking care of runtime values.

tests:
- deploy a config with 2 weighted cluster weight 1/100
```
  "routing.cluster_weight.fooA": {
   "final_value": "1",
   "layer_values": [
    "1",
    "",
    "",
    ""
   ]
  },
  "routing.cluster_weight.fooB": {
   "final_value": "100",
   "layer_values": [
    "100",
    "",
    "",
    ""
   ]
  }
```
- deploy the new binary
- update the weight to 100/100
```
  "routing.cluster_weight.fooA": {
   "final_value": "100",
   "layer_values": [
    "100",
    "",
    "",
    ""
   ]
  },
  "routing.cluster_weight.fooB": {
   "final_value": "100",
   "layer_values": [
    "100",
    "",
    "",
    ""
   ]
  }
```
- there should be no crash

some notes/open questions:

- There is a similar function in utility:
https://github.com/envoyproxy/envoy/blob/570869520b69fcee00e3b834c40053d55907facb/source/common/common/utility.h#L505
We could refactor it to make use of it everywhere, but it is a much larger change. I believe this could be followed in another PR.

- should we get rid of `total cluster weight` parameter entirely? maybe for another PR as well? https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-weightedcluster-total-weight

this PR also fixes issue #22873

### CODE REVIEW

**Comment by wdauchy** [comment]
@ggreenway @alyssawilk is there any chance we can get someone assigned to review this PR?

**Review by tyxia — COMMENTED** [review]
Thanks for contribution! 

I agree we should fix the crash caused by hitting PANIC.  I think a few high-level things we should discuss first.

Context:
Before this PR, total weight is NOT updated by runtime.  Changing the cluster weight via runtime value makes [the selected_value](https://github.com/envoyproxy/envoy/blob/d8821370842dea5b01a458f6225b223f025f5300/source/common/router/config_impl.cc#L1376) (calculated based on config total weight) outside of the range of cluster weights(modified by runtime value)

Two ways of fixing the crash:

- (1) Re-computing the total weight at data plane, which will take account of the runtime update to total weight

- (2) Keep existing logic but return `nullptr` instead of PANIC. 

Option (1) (which is this PR's approach) seems good to me. But this will be a behavior/api change that is probably good to get @envoyproxy/api-shepherds 's input.

Besides the high-level direction of this PR:

- (1) For option 1, can we limit the re-computation of total weight when [runtime_key_prefix](https://github.com/envoyproxy/envoy/blob/d2a20a02b6664c531e51cce04cc4283cb554ed4c/api/envoy/config/route/v3/route_components.proto#L489) is configured. 
    This re-computation introduces data plane overhead which is redundant when there is no runtime update.

- (2) can you also add a test that originally will hit PANIC code. You can update it to a very small new weight range.

**Comment by wdauchy** [comment]
Hi @tyxia 

Thanks for your review and sorry for the delay to address the comments.

> * (1) For option 1, can we limit the re-computation of total weight when [runtime_key_prefix](https://github.com/envoyproxy/envoy/blob/d2a20a02b6664c531e51cce04cc4283cb554ed4c/api/envoy/config/route/v3/route_components.proto#L489) is configured.
>   This re-computation introduces data plane overhead which is redundant when there is no runtime update.
> * (2) can you also add a test that originally will hit PANIC code. You can update it to a very small new weight range to repro this

I believe I addressed both. It makes some aspects slightly harder to read as we have to condition on `runtime_key`. Let me know it is good enough for you.

**Comment by tyxia** [comment]
Before looking into code change, i think it will be good to have another API shepherd review the on the [proposed direction of the fix](https://github.com/envoyproxy/envoy/pull/34793#pullrequestreview-2136283182)

@adisuissa I know you have worked on weighted cluster divide-by-0 crash before, could you also take a look of direction above? Thanks!

**Comment by wdauchy** [comment]
@adisuissa let me know if I can clarify something

**Comment by adisuissa** [comment]
> (1) For option 1, can we limit the re-computation of total weight when [runtime_key_prefix](https://github.com/envoyproxy/envoy/blob/d2a20a02b6664c531e51cce04cc4283cb554ed4c/api/envoy/config/route/v3/route_components.proto#L489) is configured.
This re-computation introduces data plane overhead which is redundant when there is no runtime update.

I agree that recomputing the total-weight only when the runtime_key_prefix is enabled makes sense.

Regarding whether to return nullptr vs PANIC, I think it would be better to keep the PANIC.
If nullptr is returned, it requires all callers to handle that, otherwise the issue is just punted up the call stack, and makes it harder to debug. If there are other cases that cause a PANIC, they should be handled explicitly in this code.
In the context of this change:
- if the total-weight is 0 because no cluster doesn't exist - I wonder if this case can even happen.
- if there's overflow of the sum of all weights - would using the original weights make sense here (with an appropriate warning)?

**Comment by wdauchy** [comment]
Thanks for your reply @adisuissa 

> In the context of this change:
> 
> * if the total-weight is 0 because no cluster doesn't exist - I wonder if this case can even happen.

I think it cannot happen with regular config: https://github.com/envoyproxy/envoy/blob/b8ae3ab32964bb200894f6fc11ffe3cfbcdd9335/source/common/router/config_impl.cc#L659

but I think it is possible with runtime keys

> * if there's overflow of the sum of all weights - would using the original weights make sense here (with an appropriate warning)?

I believe this might create confusion in user's mind if we start to fallback on a different behaviour they don't expect with runtime keys. I would however propose to followup this PR by removing the deprecated `total_weight` parameter notion from the config - this would make things more logical.

**Comment by yanavlasov** [comment]
Ping @adisuissa @tyxia 

/wait-any

File: source/common/router/config_impl.cc
Comment (tyxia) [inline source/common/router/config_impl.cc:1412]: ditto.` cluster_weight`

Could you please also audit the code change and make sure it is aligned with https://github.com/envoyproxy/envoy/blob/main/STYLE.md, Thanks!

File: source/common/router/config_impl.cc
Comment (tyxia) [inline source/common/router/config_impl.cc:1378]: style nit: `absl::InlinedVector<uint32_t, 4> cluster_weights`

File: source/common/router/config_impl.cc
Comment (tyxia) [inline source/common/router/config_impl.cc:1404]: Let's discuss on error handling like here.

**Review by tyxia — COMMENTED** [review]
Yeah, we should figure out error handling of invalid runtime weight update.

My 2 cents is that we should avoid dateplane crash caused by invalid runtime weight.
@adisuissa  WDYT about `Envoy_BUG + return nullptr`? Caller-side should already be able to handle nulltpr return (as if no route match) but I do agree that returning nullptr will make debugging harder. Hopefully Envoy_BUG could help debugging here.

**Comment by adisuissa** [comment]
> My 2 cents is that we should avoid dateplane crash caused by invalid runtime weight.
@adisuissa WDYT about Envoy_BUG + return nullptr? Caller-side should already be able to handle nulltpr return (as if no route match) but I do agree that returning nullptr will make debugging harder. Hopefully Envoy_BUG could help debugging here.

I agree with ENVOY_BUG, but I'm just guessing that nullptr will crash somewhere else. IMHO it will make it a bit more challenging to debug. That said, I'm don't feel strongly about these approaches.
Maybe converting the return type to absl::StatusOr<> might be a better approach.

**Comment by tyxia** [comment]
> > My 2 cents is that we should avoid dateplane crash caused by invalid runtime weight.
> > @adisuissa WDYT about Envoy_BUG + return nullptr? Caller-side should already be able to handle nulltpr return (as if no route match) but I do agree that returning nullptr will make debugging harder. Hopefully Envoy_BUG could help debugging here.
> 
> I agree with ENVOY_BUG, but I'm just guessing that nullptr will crash somewhere else. IMHO it will make it a bit more challenging to debug. That said, I'm don't feel strongly about these approaches. Maybe converting the return type to absl::StatusOr<> might be a better approach.

AFAIU, returning nullptr here == no route found, because its caller is `RouteConstSharedPtr matches()` ---> `RouteMatcher::route`. Will envoy crash when no route found? I think it will try local rely in this case, or maybe I miss something?

**Comment by adisuissa** [comment]
> AFAIU, returning nullptr here == no route found, because its caller is `RouteConstSharedPtr matches()` ---> `RouteMatcher::route`. Will envoy crash when no route is found? I think it will local rely in this case, or maybe I miss something?

Ah, I think you are correct, and the uses of this method (that we are aware of) seem to be taking nullptr as an indication that  the route is not found. That said, previously this method did not return nullptr, so it seems a bit strange...
I guess this method should have a better/correct documentation.

I don't feel strongly about this, so I'll leave it up to you to decide.

File: source/common/router/config_impl.cc
Comment (wdauchy) [reply source/common/router/config_impl.cc]: fixed

File: source/common/router/config_impl.cc
Comment (wdauchy) [reply source/common/router/config_impl.cc]: fixed

File: source/common/router/config_impl.cc
Comment (wdauchy) [reply source/common/router/config_impl.cc]: I changed it to `ENVOY_BUG` as adviced

**Comment by wdauchy** [comment]
> Yeah, we should figure out error handling of invalid runtime weight update.

I changed it to `Envoy_BUG` as advised.

File: source/common/router/config_impl.cc
Comment (tyxia) [inline source/common/router/config_impl.cc:1412]: Thanks for updates! 

Not your code, could you also change line#1450 PANIC to IS_ENVOY_BUG to avoid dataplane crash?

File: source/common/router/config_impl.cc
Comment (tyxia) [inline source/common/router/config_impl.cc:1411]: nit: total_cluster_weight == 0

I feel that is slightly more readable but could be just my own preference.

File: source/common/router/config_impl.cc
Comment (tyxia) [inline source/common/router/config_impl.cc:1354]: Do you also want to add some documentation to this function definition, as Adi suggested https://github.com/envoyproxy/envoy/pull/34793#issuecomment-2259632156

File: source/common/router/config_impl.cc
Comment (tyxia) [inline source/common/router/config_impl.cc:1429]: nit: parentheses seems to be redundant here as It is just same as  `*cluster_weight++` due to operator ordering.
Also it is a bit confusing.

**Comment by tyxia** [comment]
/assign-from @envoyproxy/senior-maintainers 

This PR touches Envoy core code.

**Review by tyxia — COMMENTED** [review]
Thank you for your patience and updates!

We are getting closer. LGTM modulo some nits. I have also added senior maintainer here since this PR touches core Envoy code.

File: source/common/router/config_impl.cc
Comment (wdauchy) [reply source/common/router/config_impl.cc:1416]: fixed

File: source/common/router/config_impl.cc
Comment (wdauchy) [reply source/common/router/config_impl.cc]: fixed

File: source/common/router/config_impl.cc
Comment (wdauchy) [reply source/common/router/config_impl.cc:1358]: fixed

File: source/common/router/config_impl.cc
Comment (wdauchy) [reply source/common/router/config_impl.cc]: fixed

**Comment by wdauchy** [comment]
> We are getting closer. LGTM modulo some nits. I have also added senior maintainer here since this PR touches core Envoy code.

I addressed all the remaining comments

**Review by tyxia — DISMISSED** [review]
Nice work! Thanks for your contribution.

PS: Just a note for future, please don't use force push so that the full commit history can be kept, which makes reviewer and future reader's life easier.

**Review by yanavlasov — COMMENTED** [review]
Is it possible to add a test for the case where total cluster weight is 0 ? I see the behavior has changed but it is not clear to me returning nullptr in that case will not cause a crash.

/wait-any

**Comment by wdauchy** [comment]
> PS: Just a note for future, please don't use force push so that the full commit history can be kept, which makes reviewer and future reader's life easier.

ah sorry, I will make sure to change that in my upcoming changes

**Comment by wdauchy** [comment]
> Is it possible to add a test for the case where total cluster weight is 0 ? I see the behavior has changed but it is not clear to me returning nullptr in that case will not cause a crash.

good point, I added a test about it.

**Review by tyxia — APPROVED** [review]

**Comment by adisuissa** [comment]
@yanavlasov for a senior-maintainer review

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
